### ISSUE TITLE
API Proposal: Retrieve Mastodon Comments by URL

### ISSUE THREAD

**Issue by mofosyne** [mastodon/mastodon#29971]
### Pitch

What if we could add `GET /api/v1/comments_by_url` which would find all comments relating to a url. 
Retrieve comments associated with a specific URL or URL hash on a Mastodon instance.

```
GET /api/v1/comments_by_url
```

- `url` (string): The URL or URL hash of the page for which comments are requested.
- `limit` (integer, optional): Maximum number of comments to retrieve per request. Default is 20.
- 'moderator': Username or client side auth token of a person/org that will be filtering the comments

### Motivation

Currently such feature is mostly done via a javascript embed from Disqus which runs entirely client side (with a client side access code). But saw this original idea by https://danielpecos.com/2022/12/25/mastodon-as-comment-system-for-your-static-blog/ and tried to extend the idea so you don't have to include a self post ID. Instead I wanted to automatically search for such self post via filtering by username and url mentions in the post. This is how far I got in trying to write such function below, but found it requires authentication so that approach doesn't work.

```js
    async function findSelfPostID(mastodonInstance, accountId, postPermalinkUrl) {
      const encodedUrl = encodeURIComponent(postPermalinkUrl);
      const queryUrl = `https://${mastodonInstance}/api/v2/search?limit=1&type=statuses&account_id=${accountId}&q=${encodedUrl}`;
      console.log(`finding self post id: ${queryUrl}`);
      const response = await fetch(queryUrl);
      if (!response.ok) {
        console.log(`Cannot get self post id. Got HTTP status code: ${response.status}`);
        return null;
      }
      const data = await response.json();
      console.log(data);
      if (data.statuses.length === 0) return null;
      return data.statuses[0].id;
    }
```

So ultimately, it appears we would need to adjust the API to get something similar. In addition we should probbly add a `mastodon:` uri standard so that people can post replied more easily (by prefilling the comment with the post url)

**Comment by mofosyne** [comment]
One other consideration to think about, is that this feature if integrated into web browsers could also allow for an 'annotate-able' web.

But the main thing for the mastodon ecosystems, is that supporting embedded comments would increase the chance for non mastodon users to be exposed to mastodon comments.

**Comment by renchap** [comment]
I am not sure what you mean by "comment". Do you want to be able to retrieve every statuses that contain a given URL?

**Comment by mofosyne** [comment]
Yes that would be reasonable different approach as well (but you may need to figure out how to give site owners the option of filtering the response).

If keeping with the original approach/intent of https://danielpecos.com/2022/12/25/mastodon-as-comment-system-for-your-static-blog/ then what you would do instead is to search for the first post under an argument specified account that mentions the url. This will be the anchor point for anyone replying to it. This would also allow the siteowner to point to that mastodon post as the reply point (via a url) instead of requiring users to put the site url into their toots.

**Comment by mofosyne** [comment]
Deeper digging for related issue tickets

* [Embed Mastodon in static sites as a comment system #8316](https://github.com/mastodon/mastodon/issues/8316) - Iframe based proposal

* [Tying toots to articles for providing a unified comments infrastructure #9030](https://github.com/mastodon/mastodon/issues/9030) - This is more closer to this proposal, but is more focused on the internal architecture of mastodon. However solving that problem will make solving this issue ticket easier.
